lib/spack/llnl/util/filesystem.py

```diff
@@ -924,6 +924,71 @@ def traverse_tree(source_root, dest_root, rel_path='', **kwargs):
         yield (source_path, dest_path)
 
 
+def visit_directory_tree(root, visitor, rel_path='', depth=0):
+    """
+    Recurses the directory root depth-first through a visitor pattern
+
+    The visitor interface is as follows:
+    - handle_file(root, rel_path, depth) -> bool
+        if False is returned, the iteration over children in the current dir is stopped
+    - before_enter_dir(root, rel_path, depth) -> bool
+        if True is returned, descends into this directory
+    - before_enter_symlinked_dir(root, rel_path, depth) -> bool
+        if True is returne,d descends into this directory
+    - after_enter_dir(root, rel_path, depth) -> void
+        only called when before_enter_dir returns True
+    - after_enter_symlinked_dir(root, rel_path, depth) -> void
+        only called when before_enter_symlinked_dir returns True
+    """
+    dir = os.path.join(root, rel_path)
+
+    for f in os.listdir(dir):
+        rel_child = os.path.join(rel_path, f)
+        child = os.path.join(dir, f)
+
+        st, lst = None, None
+        is_link, is_dir = False, False
+
+        # lstat first to avoid having to stat when not a symlink
+        try:
+            lst = os.lstat(child)
+            is_link = stat.S_ISLNK(lst.st_mode)
+        except OSError:
+            pass
```

Comment:
```suggestion
            continue
```
The code below has `continue` if `lst is None and st is None`, which is already guaranteed here.